flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -195,28 +193,27 @@ void retriggerSubpartitionRequest(int subpartitionIndex) throws IOException {
 	}
 
 	@Override
-	public void spillInflightBuffers(long checkpointId, ChannelStateWriter channelStateWriter) throws IOException {
+	public void spillInflightBuffers(long checkpointId, ChannelStateWriter channelStateWriter) {
 		synchronized (receivedBuffers) {
-			checkState(checkpointId > lastRequestedCheckpointId, "Need to request the next checkpointId");
-
-			final List<Buffer> inflightBuffers = new ArrayList<>(receivedBuffers.size());
-			for (Buffer buffer : receivedBuffers) {
-				CheckpointBarrier checkpointBarrier = parseCheckpointBarrierOrNull(buffer);
-				if (checkpointBarrier != null && checkpointBarrier.getId() >= checkpointId) {
-					break;
+			final Integer numRecords = numRecordsOvertaken.remove(checkpointId);
```

Comment:
Good catch, a leak could happen when the checkpoint is cancelled through another channel. The map itself is rather small, but it could add up over all channels and gates.
I don't have a good idea on how to properly abstract this cleanup except by adding some kind of checkpoint-cancelled hook though. 
Alternatively, checkpoint barrier handler becomes more aware of the buffers to be spilled. So instead of calling `channel.spillInflightBuffers`, it could be `channel.getSpilledBuffers().forEach(channelStateWriter::write)` on a good checkpoint and `channel.getSpilledBuffers().forEach(Buffer::recycle)` on cancelled checkpoints, where `getSpilledBuffers` always cleans up this map.